apps/system/js/launcher.js

```diff
@@ -0,0 +1,108 @@
+  Launcher.SUB_MODULES = ['Scheduler'];
+  Launcher.SETTINGS = [
+    'ftu.manifestURL',
+    'homescreen.manifestURL',
+    'deviceinfo.previous_os',
+    'deviceinfo.os',
+    'lockscreen.enabled'
+  ];
+  Launcher.STATES = [
+    'justUpgraded'
+  ];
+  BaseModule.create(Launcher, {
+    name: 'Launcher',
+    DEBUG: false,
+    _start: function() {
+      this._version = {};
+      this.debug('reading configurations...');
+      return Promise.all([
+        this.readFtu(),
+        this.readSetting('homescreen.manifestURL'),
+        this.readSetting('ftu.manifestURL'),
+        this.readSetting('deviceinfo.os'),
+        this.readSetting('deviceinfo.previous_os'),
+        this.readSetting('lockscreen.enabled')
+      ]).then(function(results) {
+        var [ftuEnabled, homescreenManifestURL, ftuManifestURL,
+             osVersion, previousOsVersion, lockscreenEnabled] = results;
+        this.lockscreenEnabled = lockscreenEnabled;
+        this.debug('Ftu enabled = ' + ftuEnabled +
+          '; ftuManifestURL = ' + ftuManifestURL +
+          '; osVersion = ' + osVersion +
+          '; previousOsVersion = ' + previousOsVersion +
+          '; homescreenManifestURL = ' + homescreenManifestURL +
+          '; lockscreenEnabled = ' + lockscreenEnabled);
+        if (this.checkUpgrading(osVersion, previousOsVersion)) {
+          this.debug('upgrading boot');
+          this.launchFtuThenHomescreen(ftuManifestURL, homescreenManifestURL);
+        } else {
+          if (ftuEnabled !== false && ftuManifestURL) {
+            this.debug('clean boot');
+            this.launchFtuThenHomescreen(ftuManifestURL, homescreenManifestURL);
+          } else {
+            this.debug('normal boot');
+            // We still need to tell FtuLauncher to skip to process some tasks.
+            this.service.request('FtuLauncher:skip');
+            if (lockscreenEnabled) {
+              // XXX: We should have a lockscreenLauncher here.
+              this.service.request('LockScreenLauncher:launch').then(() => {
+                this.service.request('LogoManager:animatePoweronLogo');
+                this.service.request('HomescreenLauncher:launch',
+                  homescreenManifestURL);
+              });
+            } else {
+              this.service.request('HomescreenLauncher:launch',
+                homescreenManifestURL).then(() => {
+                  this.service.request('LogoManager:animatePoweronLogo');
+                });
+              this.service.request('LockScreenLauncher:standby');
+            }
+          }
+        }
+      }.bind(this));
+    },
+    justUpgraded: function() {
+      return this._isUpgrade;
+    },
+    checkUpgrading: function(prev, curr) {
+      var isUpgrade = false;
+      // dont treat lack of previous version info as an upgrade
+      if (prev && curr) {
+        isUpgrade = curr.major > prev.major || curr.minor > prev.minor;
+      }
+      this._isUpgrade = isUpgrade;
```

Comment:
"BaseModule is a class skeleton which helps you to build a module with (such abilities)" --> I know, from the way you use it and the comment like, your `BaseModule.create` is generate a `class`, so it would be like `react.class` or `backbone.extend`.  I just wonder if it's necessary to give a literal object to do the mixin (or, what's wrong with the old plain `Class.prototype.method`), or we could give it any object so that I could still feed a defined `prototype` object to do that, which for me is better than create an odd `stateful literal object`.